pkg/controller/autoscaling/elasticsearch/controller.go

```diff
@@ -0,0 +1,228 @@
+
+const (
+	controllerName = "elasticsearch-autoscaling"
+
+	enterpriseFeaturesDisabledMsg = "Autoscaling is an enterprise feature. Enterprise features are disabled"
+)
+
+var defaultReconcile = reconcile.Result{
+	Requeue:      true,
+	RequeueAfter: 60 * time.Second,
+}
+
+// ReconcileElasticsearch reconciles autoscaling policies and Elasticsearch resources specifications based on autoscaling decisions.
+type ReconcileElasticsearch struct {
+	k8s.Client
+	operator.Parameters
+	esClientProvider EsClientProvider
+	recorder         record.EventRecorder
+	licenseChecker   license.Checker
+
+	// iteration is the number of times this controller has run its Reconcile method
+	iteration uint64
+}
+
+// NewReconciler returns a new reconcile.Reconciler
+func NewReconciler(mgr manager.Manager, params operator.Parameters) *ReconcileElasticsearch {
+	c := mgr.GetClient()
+	return &ReconcileElasticsearch{
+		Client:           c,
+		Parameters:       params,
+		esClientProvider: newElasticsearchClient,
+		recorder:         mgr.GetEventRecorderFor(controllerName),
+		licenseChecker:   license.NewLicenseChecker(c, params.OperatorNamespace),
+	}
+}
+
+// Reconcile updates the ResourceRequirements and PersistentVolumeClaim fields for each elasticsearch container in a
+// NodeSet managed by an autoscaling policy. ResourceRequirements are updated according to the response of the Elasticsearch
+// _autoscaling/capacity API and given the constraints provided by the user in the autoscaling specification.
+func (r *ReconcileElasticsearch) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	ctx = common.NewReconciliationContext(ctx, &r.iteration, r.Tracer, controllerName, "es_name", request)
+	defer common.LogReconciliationRunNoSideEffects(logconf.FromContext(ctx))()
+	defer tracing.EndContextTransaction(ctx)
+
+	// Fetch the Elasticsearch instance
+	var es esv1.Elasticsearch
+	requeue, err := r.fetchElasticsearch(ctx, request, &es)
+	if err != nil || requeue {
+		return reconcile.Result{}, tracing.CaptureError(ctx, err)
+	}
+
+	if !es.IsAutoscalingDefined() {
+		return reconcile.Result{}, nil
+	}
+
+	log := logconf.FromContext(ctx)
+
+	enabled, err := r.licenseChecker.EnterpriseFeaturesEnabled()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if !enabled {
+		log.Info(enterpriseFeaturesDisabledMsg)
+		r.recorder.Eventf(&es, corev1.EventTypeWarning, license.EventInvalidLicense, enterpriseFeaturesDisabledMsg)
+		// We still schedule a reconciliation in case a valid license is applied later
+		return defaultReconcile, nil
+	}
+
+	if common.IsUnmanaged(&es) {
+		log.Info("Object is currently not managed by this controller. Skipping reconciliation", "namespace", es.Namespace, "es_name", es.Name)
+		return reconcile.Result{}, nil
+	}
+
+	selector := map[string]string{label.ClusterNameLabelName: es.Name}
+	compat, err := annotation.ReconcileCompatibility(ctx, r.Client, &es, selector, r.OperatorInfo.BuildInfo.Version)
```

Comment:
Good point, since I think we still want to check the compatibility I introduced a dedicated function in https://github.com/elastic/cloud-on-k8s/pull/4197